src/sdl2/video.rs

```diff
@@ -885,38 +898,43 @@ impl WindowBuilder {
 
 impl Window {
     #[inline]
-    pub fn raw(&self) -> *mut ll::SDL_Window { self.raw }
+    pub fn raw(&self) -> *mut ll::SDL_Window { self.context.raw }
 
     #[inline]
     pub unsafe fn from_ll(subsystem: VideoSubsystem, raw: *mut ll::SDL_Window) -> Window {
-        Window {
-            subsystem: subsystem,
-            raw: raw
-        }
+        let context = WindowContext {
+            subsystem: subsystem.clone(),
+            raw: raw,
+        };
+        Window { context: Rc::new(context) }
     }
 
     #[inline]
-    pub fn subsystem(&self) -> &VideoSubsystem { &self.subsystem }
+    pub fn subsystem(&self) -> &VideoSubsystem { &self.context.subsystem }
 
     /// Initializes a new `RendererBuilder`; a convenience method that calls `RendererBuilder::new()`.
     pub fn renderer(self) -> RendererBuilder {
```

Comment:
Rename into `into_canvas` please, but keep the Builder (just rename the builder into `CanvasBuilder`).
If possible, make it so that you can have a CanvasBuilder both from `let canvas : Result<Canvas<_>,_> = Window.into_canvas().build()` and `let canvas : Result<Canvas<_>,_> = CanvasBuilder::from_window(window).build()`. Just make one alias the other or something.